### ISSUE TITLE
Experimental conda packages for ifcopenshell v0.6.0 dev branch

### ISSUE THREAD

**Issue by tpaviot** [IfcOpenShell/IfcOpenShell#1327]
I built experimental conda packages for ifcopenshell branch 0.6.0 latest commit. It will be easier for everyone to follow the most recent developments and test the package.

It is available from my own anaconda channel:
```bash
conda install -c tpaviot -c conda-forge ifcopenshell==0.6.0.a0f79cc
```

(the commit number is appended as a suffix to the branch name to build the version identifier). It is currently built with most options (but not the Collada support), against opencascade-7.4.0 (the one from the conda-forge channel), for python 3.7 and 3.8 Windows/Linux/OSX. pythonocc-core 7.4.1 (conda-forge) is set as a dependency and is installed as well.

Conda recipe: https://github.com/tpaviot/IfcOpenShell/blob/azure-ci/conda/

MS Azure pipeline: https://dev.azure.com/tpaviot/ifcopenshell/_build?definitionId=7

Anaconda channel: https://anaconda.org/tpaviot/ifcopenshell

The workflow I set up is the following:
myIfcOpenShell github fork commit -> MS Azure compilation farm Lin/OSX/Win -> Anaconda.org repository.

This is then quite straightforward to stay up-to-date with IfcOpenShell v0.6.0 branch. So far, there are *many* commits on my fork, I wait for the packages to be tested before pushing a pull request to contribute a clean conda recipe. There are also some required changes to the CMake file.

What I expect from the community is to:

*  report success/failures/issues (especially on OSX, I don't have any mac machine to test from a fresh conda local installation),

*  suggest the best timestep to provide up-to-date packages (daily is too much work, maybe weekly? monthly? provide packages for tags only with monthly tags? etc.

* provide any suggestion/fix/improvement, as usual.

**Comment by WNepomuk** [comment]
Is there anything new on this? I really like the possibility to install ifcopenshell using Conda. It seems like the Conda Channel didn't get updated in the last months. It would be great to have Conda Packages with the recent changes if at all possible.

**Comment by aothms** [comment]
For now this lives on @tpaviot 's fork, which is 1000 commits behind. This explains the lack of pushes to conda.

There are some clear open issues in the OP that didn't get answered:

- *Testing especially OSX*. I assume @WNepomuk did this and didn't find issues, so that's good?
- *Required changes to the CMake file*. @tpaviot can you elaborate on this?
- *Best timestep to provide up-to-date packages.* This is rather crucial. IfcOpenShell doesn't really have tags and branches stay active for years. For now I would just suggest weekly. I know Github Actions has a cron like syntax https://docs.github.com/en/actions/reference/events-that-trigger-workflows#scheduled-events Azure seems to have as well https://docs.microsoft.com/en-us/azure/devops/pipelines/process/scheduled-triggers?view=azure-devops&tabs=yaml#examples
- *Question 1 from my side*. We have now three channels. What are the benefits of a custom workflow one vs conda-forge?
    1. manual ifcopenshell (I think it can be discarded)
    2. conda-forge
    3. Azure CI from @tpaviot 
- *Question 2 from my side*. I'm a bit reluctant on adding another CI provider. We moved away from Travis and have quite few Github Actions workflows. What are your thoughts @tpaviot . The differences between Azure and Github Actions seem small (but the devil is in the details of course, I didnt take a thorough look) https://docs.github.com/en/actions/learn-github-actions/migrating-from-azure-pipelines-to-github-actions They are both owned by microsoft. Which of these two platforms is going to stay?

**Comment by tpaviot** [comment]
Thank you @WNepomuk for the follow-up regarding this issue. I did not work on it for a while, the repository needs an update but I have currently little free time.

Question 1 @aothms: conda-forge is perfect for delivery, Azure-CI for continuous integration. Using Azure CI, you can run unit tests, benchmarks, doc building etc., all sorts of things that are not in the scope of conda-forge. With conda-forge, you can build your projects and upload packages to the anaconda cloud. With Azure CI, you can do both: CI and CD, that's why I prefer this solution over conda-forge only. Moreover, the conda-forge workflow is a bit weird.

Question 2. I did not set up github actions on my personal projects. I have the feeling that gh actions are just a way to trigger builds on Azure CI and get the results. According to me, gh actions allow using Azure CI within github transarently, while Azure CI has a broader audience (no need to be on gthub to use Azure CI). Honestly, I don't see what the gh actions bring over a classical azure workflow I should carefully read the migration documentation.

**Comment by aothms** [comment]
Makes sense @tpaviot I think we can merge this as is and maybe later somebody feels like unifying all the CI/CD scripts to Github Actions. A PR at your earliest convenience would be greatly appreciated :) if there are rough edges that's fine just let me know if something can still be improved and I'll try to work on it.

**Comment by tpaviot** [comment]
Ok, let me update my local copy to the latest head and try to rebase my old work on top of it, merge commit in order to make a clean PR. I realize that many changes to the CMakeList.txt are not necessary and not related to this issue, I'll try to have something less intrusive

**Comment by aothms** [comment]
@tpaviot  Wonderful!

**Comment by tpaviot** [comment]
I updated my ```azure-ci``` branch to be on top of latest ``v0.6.0``.

@aothms it seems previous cmake changes are not required for the build to complete, but a lot of warnings are raised. Which occt version should be preferred (7.4.0? 7.5.0? 7.5.1?). conda build currently relies on occt-7.4.0 (7.5.x fails to build on Linux)

@WNepomuk linux and osx packages are already available on my anaconda channel, windows binaries are on the way (windows builds use to take hours). Can you please test one of these packages and report the result?

**Comment by aothms** [comment]
> Which occt version should be preferred (7.4.0? 7.5.0? 7.5.1?). conda build currently relies on occt-7.4.0 (7.5.x fails to build on Linux)

Both 7.3 and 7.5 have quite a bit fewer issues than 7.4 when ran with IfcOpenShell, but for day to day usage there shouldn't be major differences.

> 7.5.x fails to build on Linux

What's the error? I use 7.5.1 as the default OCCT for the ifcopenshell 0.7 branch and didn't get any issues. But I compile static libs, perhaps there is a difference there.

**Comment by WNepomuk** [comment]
> @WNepomuk linux and osx packages are already available on my anaconda channel, windows binaries are on the way (windows builds use to take hours). Can you please test one of these packages and report the result?

Thank you very much! That's perfect :)
I will check it out (on mac OS) and let you know how it went.

**Comment by tpaviot** [comment]
@aothms ok I will go with 7.5. I should have said the conda build fails (library conflicts between glibc/libxml2/occt etc).  On my machine that works fine. conda packages come with a lot of dependencies (especially conda-forge packages) and, thus, with possible version conflicts.

**Comment by WNepomuk** [comment]
I'm back with the result of my testing. The installation did not seem to work for me on either MacOS, Linux or Windows. Since I am not an expert on python and conda I'm not sure if I did everything correctly. So please let me know if I did something wrong and/or should try something different.

What I tried so far:

## Mac OS

First I created a new conda environment using python 3.9 because I need to use this version for a project. I tried to install ifcopenshell there:

```
(testenv) Nepomuks-Air:~ Nepomuk$ conda install -c tpaviot ifcopenshell
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: \ 
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed                                                                                                                                                       

UnsatisfiableError: The following specifications were found
to be incompatible with the existing python installation in your environment:

Specifications:

  - ifcopenshell -> python[version='>=3.7,<3.8.0a0|>=3.8,<3.9.0a0']

Your python: python=3.9

If python is on the left-most side of the chain, that's the version you've asked for.
When python appears to the right, that indicates that the thing on the left is somehow
not available for the python version you are constrained to. Note that conda will not
change your python version to a different minor version unless you explicitly specify
that.



(testenv) Nepomuks-Air:~ Nepomuk$ 
```
Since it seems to need python 3.7 or 3.8 I created a new conda environment with pyhton 3.8 and tried installing ifcopenshell:

```
(testenv) Nepomuks-Air:~ Nepomuk$ conda install -c tpaviot ifcopenshell
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: \ 
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed                                                                                                                                                       

UnsatisfiableError: The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions

(testenv) Nepomuks-Air:~ Nepomuk$
```
These are the packages in the conda environment:
```
(testenv) Nepomuks-Air:~ Nepomuk$ conda list
# packages in environment at /Users/Nepomuk/opt/anaconda3/envs/testenv:
#
# Name                    Version                   Build  Channel
ca-certificates           2021.7.5             hecd8cb5_1  
certifi                   2021.5.30        py38hecd8cb5_0  
libcxx                    12.0.0               h2f01273_0  
libffi                    3.3                  hb1e8313_2  
ncurses                   6.2                  h0a44026_1  
openssl                   1.1.1l               h9ed2024_0  
pip                       21.0.1           py38hecd8cb5_0  
python                    3.8.11               h88f2d9e_1  
readline                  8.1                  h9ed2024_0  
setuptools                52.0.0           py38hecd8cb5_0  
sqlite                    3.36.0               hce871da_0  
tk                        8.6.10               hb0a8c7a_0  
wheel                     0.37.0             pyhd3eb1b0_1  
xz                        5.2.5                h1de35cc_0  
zlib                      1.2.11               h1de35cc_3  
(testenv) Nepomuks-Air:~ Nepomuk$
```
Using python 3.7 gave the same result as using python 3.8.   
Finally I created a new empty conda environment (without python installed) and tried to install ifcopenshell, which gave me the following error:
```
(testenv) Nepomuks-Air:~ Nepomuk$ conda install -c tpaviot ifcopenshell
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: / 
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed                                                                                                                                                      

UnsatisfiableError: 

(testenv) Nepomuks-Air:~ Nepomuk$
```
I asked someone to test the installation on Linux and Windows as well. In both cases we did create a new conda environment and tried to install ifcopenshell:  
## Linux
```
(slab) boxy@7c5fa7c9429d:~$  cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

(slab) boxy@7c5fa7c9429d:~$ conda install -c tpaviot ifcopenshell
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: \
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed

UnsatisfiableError: The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versionsThe following specifications were found to be incompatible with your system:

  - feature:/linux-64::__glibc==2.31=0
  - feature:|@/linux-64::__glibc==2.31=0
  - ifcopenshell -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']

Your installed version is: 2.31
```

## Windows

```
(slab21) PS C:\Users\felix> conda install -c tpaviot ifcopenshell
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: /
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed

UnsatisfiableError: The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions
```

I hope this helps a little bit. Should I try something else?

**Comment by aothms** [comment]
OSX

~~~
Specifications:

  - ifcopenshell -> python[version='>=3.7,<3.8.0a0|>=3.8,<3.9.0a0']

Your python: python=3.9
~~~

This means your python version is not available in the channel. This is something I encounter rather frequently on conda. You can solve this by downgrading your python version `conda install python=3.8`

Linux

~~~
  - feature:/linux-64::__glibc==2.31=0
  - feature:|@/linux-64::__glibc==2.31=0
  - ifcopenshell -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']
~~~

This I do not understand. It seems to me that 2.31 > 2.17. Which should be ok.

Windows

The Windows one is not detailed enough to comment. Is there more output?

---

So it seems at least the OSX one is easily solvable by downgrading python. Tbh this is quite in line with my daily experience of using conda. That's why I kind of gave up on it and just publish standalone zip packages on the ifcopenshell website. It causes other problems, sure, and you can still have incompatible glibc versions, but it's more overseeable for me.

**Comment by tpaviot** [comment]
@WNepomuk it's not ready yet. Windows and osx builds suffer from linkage issues with python library. Let me notify you when it is fixed.

**Comment by tpaviot** [comment]
@WNepomuk Can you please test one more time? I updated the build process, everything should work now (packages available for win/linux/osx for py 3.7/3.8 and 3.9)

**Comment by WNepomuk** [comment]
I had problems again at first, but after I added the channels ```conda-forge```, ```oce``` and  ```dlr-sc``` to the conda environment by running ```conda config --add channels <channel>```it installed ifcopenshell correctly.
I added these channels because they are mentioned in the installation guide on the ifcopenshell github. I am not sure if all the channels are needed, but as far as I can tell everything works properly now. I tried it for python versions 3.7, 3.8 and 3.9 and it worked for each one of them.
I am on MacOS right now but will try to install it on windows and linux when I get a chance as well. 

Thank you very much for your work and time, ifcopenshell is an awesome tool to have!

**Comment by tpaviot** [comment]
@WNepomuk There's no need to add the oce and dlr-sc channels. Only tpaviot and conda-forge are required, any other additional package may create library conflicts.

@aothms Before I create a PR, I have a few questions:
* what kind of versioning convention would you prefer? Currently, the package id is ifcopenshell-0.6.0, would you like the git number appended as a suffix, e.g. ifcopenshell-0.6.0.f2362e13?
* The build workflow is quite long (up to 3 hours for Windows). I don't think it's worth triggering a build for each commit, maybe a nightly build is enough, what do you think?
* the packages are uploaded to my personal channel. I can leave the code as is, create the PR, and after that your change the code to upload the package to another conda channel, wherever you want. Or I can create another organization, named ifcopenshell, where the packages would be uploaded, and take care of it. Your opinion?

**Comment by tpaviot** [comment]
@aothms I didn't see an ifcopenshell is already there, no need to create another one then

**Comment by aothms** [comment]
> would you like the git number appended as a suffix, e.g. ifcopenshell-0.6.0.f2362e13?

I think that'd be great. I don't fully oversee the consequences though. Does it mean that users can install with `install ifcopenshell==0.6`? That'd be ideal as 0.7 is coming up as something stable.

> I don't think it's worth triggering a build for each commit, maybe a nightly build is enough, what do you think?

I would say weekly. Most changes these days are blenderbim which has it's own release flow.

> I didn't see an ifcopenshell is already there, no need to create another one then

yes exactly. it's horribly outdated so your PR is highly appreciated. That means that the install command will be `install -c ifcopenshell -c tpaviot -c conda-forge ifcopenshell?`

**Comment by tpaviot** [comment]
> > would you like the git number appended as a suffix, e.g. ifcopenshell-0.6.0.f2362e13?
> 
> I think that'd be great. I don't fully oversee the consequences though. Does it mean that users can install with `install ifcopenshell==0.6`? That'd be ideal as 0.7 is coming up as something stable.

No, if the git number is appended to the version number, it won't be possible to install ifcopenshell==0.6.0, you will have to specify the git number as well, e.g. install ifcopenshell==0.6.0.bbccddeeff If the git commit number is appended as a suffix, then it will be possible to install *any* git commit, and roll back to a previous version in case of a regression. If ever the package version is only 0.6.0, then each new package overloads the older one. I would go with 0.6.0 only, no git number.
 
> 
> > I don't think it's worth triggering a build for each commit, maybe a nightly build is enough, what do you think?
> 
> I would say weekly. Most changes these days are blenderbim which has it's own release flow.
> 

ok, that makes sense.

> > I didn't see an ifcopenshell is already there, no need to create another one then
> 
> yes exactly. it's horribly outdated so your PR is highly appreciated. That means that the install command will be `install -c ifcopenshell -c tpaviot -c conda-forge ifcopenshell?`

No, that would be ``install -c conda-forge -c ifcopenshelll ifcopenshell==0.6.0``.

**Comment by aothms** [comment]
> I would go with 0.6.0 only, no git number.

I agree then. Commit sha is too much detail. 

Looking forward to the PR.

**Comment by aothms** [comment]
@tpaviot any updates :) ?

**Comment by luzpaz** [comment]
bump

**Comment by Moult** [comment]
Given that a lot of work is now on v0.7.0, is it time to revisit Anaconda packaging for v0.7.0?

**Comment by Moult** [comment]
This can be closed due to #2132 :) Hooray!
